scene/animation/animation_tree.h

```diff
@@ -380,6 +390,37 @@ class AnimationTree : public Node {
 	Variant post_process_key_value(const Ref<Animation> &p_anim, int p_track, Variant p_value, const Object *p_object, int p_object_idx = -1);
 	virtual Variant _post_process_key_value(const Ref<Animation> &p_anim, int p_track, Variant p_value, const Object *p_object, int p_object_idx = -1);
 
+	struct Record {
+		List<AnimationNode::AnimationState> state_animation_states;
+		uint64_t state_last_pass;
+		bool state_valid;
+		String state_invalid_reasons; // Optional
+
+		uint64_t process_pass;
+
+		Vector3 root_motion_position;
+		Quaternion root_motion_rotation;
+		Vector3 root_motion_scale;
+		Vector3 root_motion_position_accumulator;
+		Quaternion root_motion_rotation_accumulator;
+		Vector3 root_motion_scale_accumulator;
+
+		HashMap<StringName, Pair<Variant, bool>> property_map;
+
+		// int state_track_count = 0; // Will be change when update chaches
+		// HashMap<NodePath, int> state_track_map; // Will be change when update chaches
+		// AnimationPlayer *state_player = nullptr; // Will be updated in _process_graph(), save it is not make sense.
+		// AnimationTree *state_tree = nullptr; // This pointer, not need to save.
+		// bool started; // Save only occur after _process_graph() and always equal false.
+	};
+
+	uint32_t record_buffer_max_size = 1024;
+	List<Record> record_buffer;
```

Comment:
Should probably be a `Vector` or `LocalVector` since it is accessed by random access

Reply:
The ideal type should be `CircalQueue` with random accessing, but godot has not this template.

When the buffer reach max size and pop the oldest record, `Vector` will lead to move all elements.

In rollback use case, store state in buffer is usually than read from buffer, `List` maybe better than `Vector`.

Reply:
I create a new  template `CircularDeque` to handle this use case.
I have not do test about this, but it should be more efficient.